### ISSUE TITLE
Multiple Joins + Nested Joins cannot be used in combination

### ISSUE THREAD

**Issue by sawirricardo** [rakibtg/SleekDB#104]
Hi, @rakibtg , first of all, I want to thank you. Your work is amazing and really love it. Useful for mocking database before diving deep into mysql.

I found an issue where I cannot make aggregation query multiple join + nested join.

To give you the idea, this code will only query multiple joins, but not the nested join.
```
$arr_user = $db_users->createQueryBuilder()
     ->join(function ($arr_user) use ($db_posts) {
         return $db_posts->findBy(['user_id', '=', $arr_user['_id']]);
     }, 'posts')
        ->join(function ($arr_user) use ($db_comments, $db_posts) {
            return $db_comments
                ->createQueryBuilder()
                ->where(['user_id', '=', $arr_user['_id']])
                ->join(function ($arr_comment) use ($db_posts) {
                    return $db_posts->findById($arr_comment['thread_id']); //<---- This query will not work
                }, 'post')->getQuery()->first();
        }, 'comments')
        ->getQuery()->first();
```

If I take out the multiple joins, the nested join works, like in the example.
```
$arr_user = $db_users->createQueryBuilder()
        ->join(function ($arr_user) use ($db_comments, $db_posts) {
            return $db_comments
                ->createQueryBuilder()
                ->where(['user_id', '=', $arr_user['_id']])
                ->join(function ($arr_comment) use ($db_posts) {
                    return $db_posts->findById($arr_comment['thread_id']);
                }, 'post')->getQuery()->first();
        }, 'comments')
        ->getQuery()->first();
```

Do you have any ideas? Thank you 👍🏻

**Comment by sawirricardo** [comment]
Oops sorry, I think the joins have worked again like charm. The issue is because I didn't specify "where", yet the query is also valid which make the code didn't output any errors.

If you cannot make this works, please double check your query is correct. It may not output any errors.

**Comment by sawirricardo** [comment]
Hi, really sorry for the confusion, but on second observation, there are something strange with this multiple and nested combination.

**Comment by Timu57** [comment]
Hi @sawirricardo 
can you tell me what result you got and what you expected?
It also would be nice to know what kind of errors are thrown..
Thank you in advance

**Comment by sawirricardo** [comment]
> Hi @sawirricardo
> can you tell me what result you got and what you expected?
> It would also be nice to know what kind of errors are thrown..
> Thank you in advance

Hi @Timu57 For the multiple joins and nested joins, I expect it would look like this
```
username => ''
posts => [
              [ 'post_id'=>1],
              [ 'post_id'=>2],
],
comments => [
              [
                'comment_id' => 1,
                'post' => [
                           'post_id' => 1,
                 ]
              ],
              [
                'comment_id' => 2,
                'post' => [
                           'post_id' => 2,
                 ]
              ],
          ],
```

However the result would look like this
```
username => ''
post => [ 'post_id'=>1],
comments => [
               [ 'comment_id' => 1],
               [ 'comment_id' => 1], // <- no post field
          ],
```

And the thing was I didn't get any errors. What I did was to try take out the multiple joins, and thus the nested joins works. If I takeout the nested joins, then the multiple joins can work.

Although it is solvable by just querying 1 by 1, I think it will worth to take a look to this issue.

**Comment by rakibtg** [comment]
Hi @sawirricardo, 

What is the updated query method looks like?

**Comment by Timu57** [comment]
@sawirricardo I am sorry but I could not reproduce the issue.

I used the following code:

```php
<?php
// index.php

require_once './vendor/autoload.php';

use SleekDB\Store;

$userStore = new Store('users', __DIR__ . '/data');
$postStore = new Store("posts", __DIR__ . "/data");
$commentStore = new Store("comments", __DIR__ . "/data");

//    $users = [
//        ["name" => "Foo", "age" => 10],
//        ["name" => "Bar", "age" => 20],
//        ["name" => "Fizz", "age" => 30],
//        ["name" => "Baz", "age" => 40],
//    ];
// $userStore->insertMany($users);

//    $posts = [
//        ["user_id" => 1],
//        ["user_id" => 1],
//        ["user_id" => 2],
//        ["user_id" => 3],
//        ["user_id" => 3],
//    ];
//    $postStore->insertMany($posts);

//    $comments = [
//        ["user_id" => 2, "post_id" => 2],
//        ["user_id" => 3, "post_id" => 3],
//        ["user_id" => 3, "post_id" => 1],
//        ["user_id" => 1, "post_id" => 4],
//        ["user_id" => 1, "post_id" => 1],
//    ];
//    $commentStore->insertMany($comments);

$result = $userStore
  ->createQueryBuilder()
  ->join(function ($user) use ($postStore){
    return $postStore->findBy(["user_id", "=", $user["_id"]]);
  }, "posts")
  ->join(function($user) use ($commentStore, $postStore){
    return $commentStore
      ->createQueryBuilder()
      ->where(["user_id", "=", $user["_id"]])
      ->join(function ($comment) use ($postStore){
        return $postStore->findById($comment["_id"]);
      }, "posts")
      ->getQuery()
      ->first();
  }, "comments")
  ->getQuery()
  ->fetch();

header("Content-Type: application/json");

echo json_encode($result);

```

```json
[
  {
    "name":"Fizz",
    "age":30,
    "_id":3,
    "posts": [
      {
        "user_id":3,
        "_id":9
      },
      {
        "user_id":3,
        "_id":10
      },
      {
        "user_id":3,
        "_id":5
      },
      {
        "user_id":3,
        "_id":4
      }
    ],
    "comments": {
      "user_id":3,
      "post_id":1,
      "_id":3,
      "posts": {
        "user_id":2,
        "_id":3
      }
    }
  },
  {
    "name":"Bar",
    "age":20,
    "_id":2,
    "posts": [
      {
        "user_id":2,
        "_id":8
      },
      {
        "user_id":2,
        "_id":3
      }
    ],
    "comments": {
      "user_id":2,
      "post_id":2,
      "_id":1,
      "posts": {
        "user_id":1,
        "_id":1
      }
    }
  },
  {
    "name":"Baz",
    "age":40,
    "_id":4,
    "posts": [],
    "comments":[]
  },
  {
    "name":"Foo",
    "age":10,
    "_id":1,
    "posts": [
      {
        "user_id":1,
        "_id":7
      },
      {
        "user_id":1,
        "_id":2
      },
      {
        "user_id":1,
        "_id":6
      },
      {
        "user_id":1,
        "_id":1
      }
    ],
    "comments": {
      "user_id":1,
      "post_id":1,
      "_id":5,
      "posts": {
        "user_id":3,
        "_id":5
      }
    }
  }
]
```

I tried to reproduce the database structure based on the query. I am not quite sure what the usecase is and why the database is structured like that but for me it looks like the query returns what is expected?

**Comment by sawirricardo** [comment]
Hi @rakibtg @Timu57 , That code quite resembles the expected result I needed. Hmm, I will try again the code tomorrow. Really appreciate the help @Timu57

**Comment by Timu57** [comment]
@sawirricardo I just noticed that the query I used was wrong.

the correct one according to your question is:

```php
<?php
// index.php

require_once './vendor/autoload.php';

use SleekDB\Store;

$userStore = new Store('users', __DIR__ . '/data');
$postStore = new Store("posts", __DIR__ . "/data");
$commentStore = new Store("comments", __DIR__ . "/data");

//    $users = [
//        ["name" => "Foo", "age" => 10],
//        ["name" => "Bar", "age" => 20],
//        ["name" => "Fizz", "age" => 30],
//        ["name" => "Baz", "age" => 40],
//    ];
// $userStore->insertMany($users);

//    $posts = [
//        ["user_id" => 1],
//        ["user_id" => 1],
//        ["user_id" => 2],
//        ["user_id" => 3],
//        ["user_id" => 3],
//    ];
//    $postStore->insertMany($posts);

//    $comments = [
//        ["user_id" => 2, "post_id" => 2],
//        ["user_id" => 3, "post_id" => 3],
//        ["user_id" => 3, "post_id" => 1],
//        ["user_id" => 1, "post_id" => 4],
//        ["user_id" => 1, "post_id" => 1],
//    ];
//    $commentStore->insertMany($comments);

$result = $userStore
  ->createQueryBuilder()
  ->join(function ($user) use ($postStore){
    return $postStore->findBy(["user_id", "=", $user["_id"]]);
  }, "posts")
  ->join(function($user) use ($commentStore, $postStore){
    return $commentStore
      ->createQueryBuilder()
      ->where(["user_id", "=", $user["_id"]])
      ->join(function ($comment) use ($postStore){
        return $postStore->findById($comment["post_id"]); // post_id not _id
      }, "posts")
      ->getQuery()
      ->first();
  }, "comments")
  ->getQuery()
  ->fetch();

header("Content-Type: application/json");

echo json_encode($result);
```

Result:
```json
[
  {
    "name": "Fizz",
    "age": 30,
    "_id": 3,
    "posts": [
      {
        "user_id": 3,
        "_id": 9
      },
      {
        "user_id": 3,
        "_id": 10
      },
      {
        "user_id": 3,
        "_id": 5
      },
      {
        "user_id": 3,
        "_id": 4
      }
    ],
    "comments": {
      "user_id": 3,
      "post_id": 1,
      "_id": 3,
      "posts": {
        "user_id": 1,
        "_id": 1
      }
    }
  },
  {
    "name": "Bar",
    "age": 20,
    "_id": 2,
    "posts": [
      {
        "user_id": 2,
        "_id": 8
      },
      {
        "user_id": 2,
        "_id": 3
      }
    ],
    "comments": {
      "user_id": 2,
      "post_id": 2,
      "_id": 1,
      "posts": {
        "user_id": 1,
        "_id": 2
      }
    }
  },
  {
    "name": "Baz",
    "age": 40,
    "_id": 4,
    "posts": [],
    "comments": []
  },
  {
    "name": "Foo",
    "age": 10,
    "_id": 1,
    "posts": [
      {
        "user_id": 1,
        "_id": 7
      },
      {
        "user_id": 1,
        "_id": 2
      },
      {
        "user_id": 1,
        "_id": 6
      },
      {
        "user_id": 1,
        "_id": 1
      }
    ],
    "comments": {
      "user_id": 1,
      "post_id": 1,
      "_id": 5,
      "posts": {
        "user_id": 1,
        "_id": 1
      }
    }
  }
]
```

It could be that something is wrong in your cache.
Try to clear it by either **deleting all files in the cache folders** of the stores ( in this example: \_\_DIR\_\_ . "/data/\<store\>/cache" where \<store\> is for example users ) **or by using the Cache class of SleekDB**:

```php
<php


require_once './vendor/autoload.php';

use SleekDB\Store;

$userStore = new Store('users', __DIR__ . '/data');
$postStore = new Store("posts", __DIR__ . "/data");
$commentStore = new Store("comments", __DIR__ . "/data");

$userStore->createQueryBuilder()->getQuery()->getCache()->deleteAll();
$postStore->createQueryBuilder()->getQuery()->getCache()->deleteAll();
$commentStore->createQueryBuilder()->getQuery()->getCache()->deleteAll();


```

**Comment by sawirricardo** [comment]
Ah I see, I think I also experienced some weird experience when developing it. (like the result not changed, or the join query is still empty, yet the query successfully return results. Do you say it may be because the caching?

**Comment by Timu57** [comment]
Sure, caching could be the problem :yum: 

I think disabling the cache functionality while development could be a good idea to avoid handling with and thinking about the caching layer while trying to find/ create the correct query.

You can disable Caching by using the ["auto_cache" configuration of the Store class](https://sleekdb.github.io/#/configurations):

```php
<?php

require_once './vendor/autoload.php';

use SleekDB\Store;

$configuration = ["auto_cache" => false];

$userStore = new Store('users', __DIR__ . '/data', $configuration);
$postStore = new Store("posts", __DIR__ . "/data", $configuration);
$commentStore = new Store("comments", __DIR__ . "/data", $configuration);
```

PS: I noticed a bug regarding the cache token generation and I will open a new issue for that.
SleekDB also uses the joins to create the cache token but there is something missing..

**Comment by sawirricardo** [comment]
Hmm I see, really appreciate the help @Timu57 . By the way, @rakibtg I didn't really notice the configuration section at first. May be you want to put a reference on the getting started section? Thank you

**Comment by Timu57** [comment]
@sawirricardo the cache token generation did not consider the join sub-queries correctly.

With the new released of [version 2.1.2](https://github.com/rakibtg/SleekDB/releases/tag/2.1.2) that problem should be fixed.

Most likely caching was the reason for your problems.

Normally the caching layer should not be a problem while development.

Thank you for your bug report.

### Update using composer
```bash
composer update rakibtg/SleekDB
```

PS: The configuration section is mentioned on the getting started page.

**Comment by sawirricardo** [comment]
> @sawirricardo the cache token generation did not consider the join sub-queries correctly.
> 
> With the new released of [version 2.1.2](https://github.com/rakibtg/SleekDB/releases/tag/2.1.2) that problem should be fixed.
> 
> Most likely caching was the reason for your problems.
> 
> Normally the caching layer should not be a problem while development.
> 
> Thank you for your bug report.
> 
> ### Update using composer
> ```shell
> composer update rakibtg/SleekDB
> ```
> 
> PS: The configuration section is mentioned on the getting started page.

Thank you for the follow-up @Timu57 👍🏻Really appreciate

**Comment by rakibtg** [comment]
Thank you @Timu57 for the fix! 
Thank you @sawirricardo for reporting the issue, we will update the doc to make it more clear about different configurations, let us know if you have any suggestions.
I will close this issue, feel free to re-open.
